extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/core/detection/DeviceRefinerImp.java

```diff
@@ -0,0 +1,149 @@
+    {
+        final EnrichmentResult.EnrichmentResultBuilder<DeviceInfoBox> resultBuilder = EnrichmentResult.builder();
+
+        final Collection<Map.Entry<String, BiPredicate<WritableDeviceInfo, DeviceInfo>>> patchPlan = buildPatchPlanFor(rawDeviceInfo);
+        if (patchPlan == null || patchPlan.isEmpty()) {
+            return resultBuilder.build();
+        }
+
+        if (populateDeviceInfo(writableAdapter, collectedEvidence, patchPlan, resultBuilder)) {
+            resultBuilder.enrichedDevice(writableAdapter.rebuildBox());
+        }
+
+        return resultBuilder.build();
+    }
+
+    private static final Map<String, PropertyMerge<WritableDeviceInfo, DeviceInfo, ?>> propertiesToMerge = Map.ofEntries(
+            entry("DeviceType", new PropertyMerge<>(DeviceInfo::getDeviceType, v -> v > 0, WritableDeviceInfo::setDeviceType)),
+            entry("Make", new PropertyMerge<>(DeviceInfo::getMake, s -> !s.isEmpty(), WritableDeviceInfo::setMake)),
+            entry("Model", new PropertyMerge<>(DeviceInfo::getModel, s -> !s.isEmpty(), WritableDeviceInfo::setModel)),
+            entry("Os", new PropertyMerge<>(DeviceInfo::getOs, s -> !s.isEmpty(), WritableDeviceInfo::setOs)),
+            entry("Osv", new PropertyMerge<>(DeviceInfo::getOsv, s -> !s.isEmpty(), WritableDeviceInfo::setOsv)),
+            entry("H", new PropertyMerge<>(DeviceInfo::getH, v -> v > 0, WritableDeviceInfo::setH)),
+            entry("W", new PropertyMerge<>(DeviceInfo::getW, v -> v > 0, WritableDeviceInfo::setW)),
+            entry("Ppi", new PropertyMerge<>(DeviceInfo::getPpi, v -> v > 0, WritableDeviceInfo::setPpi)),
+            entry("PixelRatio", new PropertyMerge<>(DeviceInfo::getPixelRatio, (BigDecimal v) -> v.intValue() > 0, WritableDeviceInfo::setPixelRatio)),
+            entry("DeviceID", new PropertyMerge<>(DeviceInfo::getDeviceId, s -> !s.isEmpty(), WritableDeviceInfo::setDeviceId))
+    );
+
+    protected Collection<Map.Entry<String, BiPredicate<WritableDeviceInfo, DeviceInfo>>> buildPatchPlanFor(
+            DeviceInfo deviceInfo
+    ) {
+        return propertiesToMerge.entrySet().stream()
+                .filter(nextMerge -> nextMerge.getValue().shouldReplacePropertyIn(deviceInfo))
+                .map(e -> new AbstractMap.SimpleEntry<String, BiPredicate<WritableDeviceInfo, DeviceInfo>>(
+                        e.getKey(),
+                        e.getValue()::copySingleValue))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    protected boolean populateDeviceInfo(
+            WritableDeviceInfo writableDeviceInfo,
+            CollectedEvidence collectedEvidence,
+            Collection<Map.Entry<String, BiPredicate<WritableDeviceInfo, DeviceInfo>>> devicePatchPlan,
+            EnrichmentResult.EnrichmentResultBuilder<?> enrichmentResultBuilder)
+    {
+        try (FlowData data = pipelineSupplier.get().createFlowData()) {
```

Comment:
so each time this method/service is called the flow data will be created, is it desired behaviour? Don't you want to read it once if it's possible?

Reply:
yes it's a requirement of the 51degrees API:  Pipeline object is reused, but FlowData has to be created for each request.

Reply:
thank you for the clarification